### PR TITLE
New version of helm chart which includes option to install manager

### PR DESCRIPTION
**Pull request by tfarinacci**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -5,3 +5,37 @@ This repository contains various [Helm](https://helm.sh) charts for installing
...
 See the readme of each chart for more information:
 * [Codesealer Ingress Sidecar Helm Chart](/charts/codesealer/README.md)
+
+## Implementation
+
+Codesealer, when implemented as a sidecar, uses an Admission Webhook to install Codesealer
+in the same pod as an Ingress Controller.  
+
+This diagram shows how the Admission Webhook comes into the picture.  Codesealer is implemented
+as a `Mutating Admission`` Webhook:
```
Comment (psve) [inline README.md]:
```suggestion
as a `Mutating Admission` Webhook:
```

File: charts/codesealer/templates/NOTES.txt [hunk 2]
```diff
@@ -1,5 +1,19 @@
+{{- if .Values.manager.enabled }}
+The first time core-manager is started, it will create a Maintenance Administrator. 
+To see the credentials for this account, check the core-manager log: 
+
+  $ kubectl get pods -n codesealer-system
+  $ kubectl logs <core-manager-pod> -n codesealer-system | grep password
+
+The manager can be accessed at https://localhost
```
Comment (psve) [inline charts/codesealer/templates/NOTES.txt]: Is this true without any port forwarding? If so, should it be?

File: charts/codesealer/templates/config-toml-configmap.yaml [hunk 3]
```diff
@@ -44,6 +48,8 @@ data:
     maxFileSize = {{ .Values.worker.config.cache.maxFileSize }}
     maxSize = {{ .Values.worker.config.cache.maxSize }}
 
+    {{- if .Values.worker.config.bootloader.fsEndpoints }}
```
Comment (psve) [inline charts/codesealer/templates/config-toml-configmap.yaml]: I'm not sure the logic is sound here. `.Values.manager.enabled` and `.Values.worker.config.bootloader.fsEndpoints` should be mutually exclusive. The easiest way to do this is to set `fsEndpoints` in this config depending on those two values, since this entire section is ignored anyway if `fsEndpoints` is false. 

We can of course still omit this section based on the value of `fsEndpoints` we actually set, if we want to make the configmap simpler when the manager is enabled.

File: charts/codesealer/templates/configmap-modsec.yaml [hunk 4]
```diff
@@ -4,23 +4,56 @@ metadata:
...
+    
+    # Include modsecurity/rules/REQUEST-903.9002-WORDPRESS-EXCLUSION-RULES.conf
+    
+    # Include modsecurity/rules/REQUEST-903.9003-NEXTCLOUD-EXCLUSION-RULES.conf
+    
+    # Include modsecurity/rules/REQUEST-903.9004-DOKUWIKI-EXCLUSION-RULES.conf
+    
+    # Include modsecurity/rules/REQUEST-903.9005-CPANEL-EXCLUSION-RULES.conf
+    
+    # Include modsecurity/rules/REQUEST-903.9006-XENFORO-EXCLUSION-RULES.conf
```
Comment (psve) [inline charts/codesealer/templates/configmap-modsec.yaml:55]: Couldn't we just omit these in this case? Same comment elsewhere in this file.

File: charts/codesealer/values.yaml [hunk 5]
```diff
@@ -228,9 +228,9 @@ worker:
       # 'TLS1.0', 'TLS1.1', 'TLS1.2', 'TLS1.3'
       # If this option is not set, TLS 1.2 is used as the default.
       tlsMinVersion: "TLS1.2"
-      # Paths for TLS certificate and key files that are used if 'network.tls' is set to true.
-      # When using the Management Portal, this certificate and key will be used as fallback
-      # when *no* certificate and key is specified in the Management Portal.
+      # Paths for TLS certificate and key files that are used if 'manager.tls' is set to true.
+      # When using the Management Portal, the certificate and key above will be used as
+      # fallback when *no* certificate and key is specified in the Management Portal.
```
Comment (psve) [inline charts/codesealer/values.yaml]: Revert this (or update to the latest wording on master).

File: charts/codesealer/values.yaml [hunk 6]
```diff
@@ -367,7 +367,7 @@ worker:
       # option, you agree to the at any time applicable TOS for Let's Encrypt. Note: The DNS
       # Record for the URL specified in 'endpoint.remoteIn' must point to Codesealer.
       letsEncrypt: false
-      # Paths for TLS certificate and key files that are used if 'network.tls' is set to true.
+      # Paths for TLS certificate and key files that are used if 'manager.tls' is set to true.
```
Comment (psve) [inline charts/codesealer/values.yaml]: Revert this.

File: tests/test_install.sh [hunk 7]
```diff
@@ -1,37 +1,39 @@
...
   echo "####################################################################################################################"
   echo "#"
-  echo "#  Usage: ./test_install.sh [install | uninstall | upgrade]"
+  echo "#  Please set CODESEALER_TOKEN variable"
   echo "#"
-  echo "####################################################################################################################"
-  exit 1
+  echo "#    $ export CODESEALER_TOKEN=<Registry Access Token>"
+  echo "#"
+  echo "#####################################################################################################################"
```
Comment (psve) [inline tests/test_install.sh]: These were swapped for a reason. Please revert.

File: tests/test_install.sh [hunk 8]
```diff
@@ -45,43 +47,19 @@ if [[ "$1" == "install" ]]; then
...
-  if [[ "${REPLY}" == 'y' ]]; then
-    helm repo add securecodebox https://charts.securecodebox.io/
+    --wait --timeout=90s
 
-    echo "########################################################################################"
-    echo "#  Waiting for Juice Shop to start"
-    echo "########################################################################################"      
+    # --set controller.updateStrategy.rollingUpdate.maxUnavailable=1 \
+    # --set controller.hostPort.enabled=true \
+    # --set controller.service.loadBalancerIP=127.0.0.1 \
```
Comment (psve) [inline tests/test_install.sh]: Why are these commented out?

File: charts/codesealer/README-standalone.md
Comment (psve) [inline charts/codesealer/README-standalone.md:1]: For the sake of `helm show readme` we shouldn't have to separate README files.

**Review by psve — COMMENTED** [review]
Some initial comments, but I haven't had time to take a closer at the docs or try it out.

File: charts/codesealer/templates/configmap-modsec.yaml (hunk 4 above)
Comment (tfarinacci) [reply charts/codesealer/templates/configmap-modsec.yaml:55]: The original file had commented rules.  I left for consistency but can remove.

File: README.md (hunk 1 above)
Comment (tfarinacci) [reply README.md]: Fixed

File: charts/codesealer/templates/NOTES.txt (hunk 2 above)
Comment (tfarinacci) [reply charts/codesealer/templates/NOTES.txt]: Yes. This works on my machine.  Seeing some Ingress issues on Felipe's machine which we are debugging.  I usually see this as http://localhost:8080 with port forwarding.  If you think that is better I can move to that.

File: charts/codesealer/templates/config-toml-configmap.yaml (hunk 3 above)
Comment (tfarinacci) [reply charts/codesealer/templates/config-toml-configmap.yaml]: Made changes to only use .Values.manager.enabled to enable standalone mode.  Let me know if you want to call this something different.

File: charts/codesealer/values.yaml (hunk 6 above)
Comment (tfarinacci) [reply charts/codesealer/values.yaml]: Done.  Consequent of copying file here.

File: charts/codesealer/values.yaml (hunk 5 above)
Comment (tfarinacci) [reply charts/codesealer/values.yaml]: Changes made

File: tests/test_install.sh (hunk 7 above)
Comment (tfarinacci) [reply tests/test_install.sh]: Done

File: tests/test_install.sh (hunk 8 above)
Comment (tfarinacci) [reply tests/test_install.sh]: Felipe was having issues on his Manjao Linux host getting the LoadBalancer to work.  It was pending on the EXTERNAL_IP when you show services.  Can we leave until we sort that out?
Also hostPort causes port conflicts when you want to run multiple workers on the same node.  I've tested the standalone with 5 workers.  Since I only have on node it was a problem.  I'd assume in prod we'd have one worker per node?

File: charts/codesealer/README-standalone.md
Comment (tfarinacci) [reply charts/codesealer/README-standalone.md:1]: Agreed and done.

File: charts/codesealer/templates/NOTES.txt (hunk 2 above)
Comment (tfarinacci) [reply charts/codesealer/templates/NOTES.txt]: I moved to http://localhost:8444

File: charts/codesealer/README.md [hunk 9]
```diff
@@ -13,33 +13,44 @@ In the following we assume that the access token is set in the following way:
...
 
 > NOTE: If using Kind, install the Ingress using the following variation
 >
 > ```bash
-> helm upgrade --install ingress-nginx ingress-nginx \
->   --repo https://kubernetes.github.io/ingress-nginx \
->   --namespace ingress-nginx --create-namespace \
+> helm repo add ${INGRESS_HELM_CHART} ${INGRESS_HELM_REPO}
+> helm install ${INGRESS_HELM_CHART} ${INGRESS_HELM_CHART}/ingress-nginx \
 >   --set controller.hostPort.enabled=true \
```
Comment (filips) [inline charts/codesealer/README.md:84]: I believe this line is also necessary for kind:

```
--set controller.service.type=NodePort 
```

File: charts/codesealer/README.md [hunk 10]
```diff
@@ -13,33 +13,44 @@ In the following we assume that the access token is set in the following way:
...
+export INGRESS_HELM_CHART=ingress-nginx
+```
+
 Additionally you will need an ingress and an application to protect. Below are steps to
 get started with a demo application and an Nginx Ingress. For guides on how to use this
 Helm chart with specific Kubernetes implementations, see the ["Kubernetes Implementation
 Specifics"](#kubernetes-implementation-specifics) section.
 
-### Ingress
-
```
Comment (filips) [inline charts/codesealer/README.md:22]: I can't get these instructions to work with minikube. I'm assuming the right way will be to use NodePort as with Kind. This, however, seems to have the side effect of disabling access to kubectl in some cases.

File: charts/codesealer/README.md [hunk 11]
```diff
@@ -13,33 +13,44 @@ In the following we assume that the access token is set in the following way:
...
+
+If you don't have an ingress already, you can install an [Nginx Ingress Controller](https://github.com/kubernetes/ingress-nginx/tree/main/charts/ingress-nginx) 
+using the following command:
 
 ```bash
-helm upgrade --install ingress-nginx ingress-nginx \
-  --repo https://kubernetes.github.io/ingress-nginx \
-  --namespace ingress-nginx --create-namespace \
-  --wait --timeout=60s
+helm repo add ${INGRESS_HELM_CHART} ${INGRESS_HELM_REPO}
```
Comment (filips) [inline charts/codesealer/README.md:68]: Why was this changed from upgrade to install?

**Review by filips — CHANGES_REQUESTED** [review]

File: charts/codesealer/README.md (hunk 10 above)
Comment (tfarinacci) [reply charts/codesealer/README.md:22]: Minikube still a work in progress.  Will go there next.

File: charts/codesealer/README.md (hunk 9 above)
Comment (tfarinacci) [reply charts/codesealer/README.md:84]: Was correct in test_install.sh.  Fixed here.

File: charts/codesealer/templates/NOTES.txt [hunk 12]
```diff
@@ -1,5 +1,27 @@
+{{- if .Values.manager.enabled }}
+The first time core-manager is started, it will create a Maintenance Administrator. 
+To see the credentials for this account, check the core-manager log: 
+
+  $ kubectl get pods -n codesealer-system
+  $ kubectl logs <core-manager-pod> -n codesealer-system | grep password
+
```
Comment (psve) [inline charts/codesealer/templates/NOTES.txt:9]: Seems like there's some text missing here?

File: charts/codesealer/templates/config-toml-configmap.yaml [hunk 13]
```diff
@@ -107,3 +117,30 @@ data:
     overrideMaxAge = {{ .Values.worker.config.endpoint.cacheRule.overrideMaxAge }}
     ignoreNoStore = {{ .Values.worker.config.endpoint.cacheRule.ignoreNoStore }}
     ignoreNoCache = {{ .Values.worker.config.endpoint.cacheRule.ignoreNoCache }}
+    {{- end }}
+    
+    {{- if .Values.manager.enabled }}
```
Comment (psve) [inline charts/codesealer/templates/config-toml-configmap.yaml]: I'd prefer having a separate config map for the manager config. I even think we have upcoming changes that will warn about undefined keys in the config, which would then happen on all workers when the manager is enabled.

File: img/sidecar-proxy.png
Comment (psve) [inline img/sidecar-proxy.png:1]: Could we please clean up this figure? At least export it from whatever drawing problem instead of taking a screenshot with controls visible.

File: README.md [hunk 14]
```diff
@@ -5,3 +5,40 @@ This repository contains various [Helm](https://helm.sh) charts for installing
...
+Codesealer, when implemented as a sidecar, uses an Admission Webhook to install Codesealer
+in the same pod as an Ingress Controller.  
+
+This diagram shows how the Admission Webhook comes into the picture.  Codesealer is implemented
+as a `Mutating Admission` Webhook:
+
+![admission webhook flow](./img/webhook.png)
+
+The `Mutating Admission` Webhook injects Codesealer as a sidecar. Traffic originally destined for the 
+Ingress Controller is `pre-routed` to Codesealer and then passed back to the Ingress Controller.  This
```
Comment (psve) [inline README.md:20]:
```suggestion
Ingress Controller is "prerouted" to Codesealer and then passed back to the Ingress Controller.  This 
```

File: README.md [hunk 15]
```diff
@@ -5,3 +5,40 @@ This repository contains various [Helm](https://helm.sh) charts for installing
...
+
+The `Mutating Admission` Webhook injects Codesealer as a sidecar. Traffic originally destined for the 
+Ingress Controller is `pre-routed` to Codesealer and then passed back to the Ingress Controller.  This 
+approach requires no changes to the application other than providing Codesealer with the ingress 
+certificates.
+
+![sidecar proxy](./img/sidecar-proxy.png)
+
+In this example the `Reverse Proxy` is Codesealer.  Traffic originally destinged for the `Ingress Controller` 
+over the ethernet eth0 interface is pre-routed to Codesealer for processing. Once Codesealer processes the payload
```
Comment (psve) [inline README.md:27]:
```suggestion
over the ethernet eth0 interface is prerouted to Codesealer for processing. Once Codesealer processes the payload 
```

File: README.md [hunk 16]
```diff
@@ -5,3 +5,40 @@ This repository contains various [Helm](https://helm.sh) charts for installing
...
+over the ethernet eth0 interface is pre-routed to Codesealer for processing. Once Codesealer processes the payload 
+it will send the traffic to the Ingress Controller over the loopback lo interface on the original port.
+
+Codesealer support multiple Ingress Controllers including NGINX Ingress, Contour, Istio Gateway,
+Kubernetes Gateway, and Cloud Provider Ingress Controllers (e.g. AWS ALB)
+
+Although this example uses an Ingress Controller for the injection, Codesealer can be injected into any pod to protect
+any service. It functions similar to an Istio Service Mesh which uses an Envoy Proxy instead.
+
+There are two methods, similar to Istio Service Mesh, to pre-route the traffic:
```
Comment (psve) [inline README.md:36]:
```suggestion
There are two methods, similar to Istio Service Mesh, to preroute the traffic:
```

File: README.md [hunk 17]
```diff
@@ -5,3 +5,40 @@ This repository contains various [Helm](https://helm.sh) charts for installing
...
+![admission webhook flow](./img/webhook.png)
+
+The `Mutating Admission` Webhook injects Codesealer as a sidecar. Traffic originally destined for the 
+Ingress Controller is `pre-routed` to Codesealer and then passed back to the Ingress Controller.  This 
+approach requires no changes to the application other than providing Codesealer with the ingress 
+certificates.
+
+![sidecar proxy](./img/sidecar-proxy.png)
+
+In this example the `Reverse Proxy` is Codesealer.  Traffic originally destinged for the `Ingress Controller`
```
Comment (psve) [inline README.md:26]:
```suggestion
In this example the `Reverse Proxy` is Codesealer.  Traffic originally destined for the `Ingress Controller` 
```

File: charts/codesealer/README.md [hunk 18]
```diff
@@ -158,7 +258,42 @@ helm repo remove codesealer
 
 ## Kubernetes Implementation Specifics
```
Comment (psve) [inline charts/codesealer/README.md:285]: The intent of this section was to outline specific instructions for this helm chart that depend on the kubernetes implementation you are using. I don't think anything added to this section in this PR belongs on a helm chart readme.

File: charts/codesealer/README.md [hunk 19]
```diff
@@ -13,32 +23,54 @@ In the following we assume that the access token is set in the following way:
...
+```bash
+export INGRESS_HELM_REPO=https://kubernetes.github.io/ingress-nginx
+export INGRESS_HELM_CHART=ingress-nginx
+export REDIS_NAMESPACE=redis
+export INGRESS_NAMESPACE=ingress-nginx
+export INGRESS_DEPLOYMENT=ingress-nginx-controller
+export INGRESS_PORT=31443
+export CODESEALER_ENV=DEVELOP
+export CODESEALER_CNI=false
+```
```
Comment (psve) [inline charts/codesealer/README.md]: The flow of this section is extremely confusing after this change. First, it says that we need to set variables related to the ingress, but then it shows variables also related to redis and codesealer. Next, it talks about a command for installing nginx ingress, but the command depends on what you've set in these variables. 

My suggestion would be to leave this section as it was (i.e., an example of what to do if you don't have an ingress. People who don't want nginx are smart enough to know what to do), and then move section with "what variables to set for this chart" down to the actual install section.

File: charts/codesealer/README.md [hunk 20]
```diff
@@ -65,9 +97,17 @@ helm install juice-shop securecodebox/juice-shop --namespace juice-shop --create
...
-Codesealer requires Redis. If you don't have your own implementation of Redis you can install
-[Bitnami package for Redis(R)](https://github.com/bitnami/charts/blob/main/bitnami/redis/README.md)
-by setting the `REDIS_NAMESPACE` variable and using the following command:
+To use this Helm chart you will also need to set the following variable to match
+your Redis deployment on your Kubernetes Cluster:
+
+```bash
+export REDIS_NAMESPACE=redis
+export REDIS_PASSWORD=<redis password>
+```
```
Comment (psve) [inline charts/codesealer/README.md]: Same here, this isn't related to installing redis, it should probably be moved to the "install" setting.

File: charts/codesealer/templates/NOTES.txt (hunk 12 above)
Comment (tfarinacci) [reply charts/codesealer/templates/NOTES.txt:9]: Does lines 2-3 provide enough context?

File: charts/codesealer/README.md (hunk 18 above)
Comment (tfarinacci) [reply charts/codesealer/README.md:285]: Removed all but Kind section

File: tests/test_install.sh (hunk 8 above)
Comment (tfarinacci) [reply tests/test_install.sh]: This is now addressed with the DEVELOP/PROD settings.

File: charts/codesealer/templates/config-toml-configmap.yaml (hunk 13 above)
Comment (tfarinacci) [reply charts/codesealer/templates/config-toml-configmap.yaml]: Completed

File: README.md (hunk 14 above)
Comment (tfarinacci) [reply README.md:20]: Fixed

File: README.md (hunk 15 above)
Comment (tfarinacci) [reply README.md:27]: Fixed

File: README.md (hunk 16 above)
Comment (tfarinacci) [reply README.md:36]: Fixed

File: README.md (hunk 17 above)
Comment (tfarinacci) [reply README.md:26]: Fixed

File: charts/codesealer/templates/NOTES.txt (hunk 12 above)
Comment (psve) [reply charts/codesealer/templates/NOTES.txt:9]: No, I was referring to the random `kubectl port-forward` in the following line. E.g. this reads like "to read the logs, create a port forward".

File: charts/codesealer/templates/image-pull-secret.yaml [hunk 21]
```diff
@@ -1,8 +1,6 @@
 apiVersion: v1
```
Comment (psve) [inline charts/codesealer/templates/image-pull-secret.yaml:1]: Should this be created when the manager is not enabled?

File: charts/codesealer/templates/ingress-netpol.yaml
Comment (psve) [inline charts/codesealer/templates/ingress-netpol.yaml:1]: What's the purpose of creating this allow all policy?

File: charts/codesealer/values.yaml [hunk 22]
```diff
@@ -297,7 +305,7 @@ worker:
       # This option can be set to 'true' in order to run the Worker in "standalone" mode. Each
       # endpoint can then be configured in this file (see the 'endpoint' section below), instead
       # of configuring them through the Management Portal.
-      fsEndpoints: true
+      # fsEndpoints: true
```
Comment (psve) [inline charts/codesealer/values.yaml]: Shouldn't this block just be removed?

File: tests/test_install.sh
Comment (psve) [inline tests/test_install.sh:1]: Honestly, I'm giving up on reviewing this file. I was okay with having a simple test script, but this is very quickly growing in complexity. It's not useful for automated testing, and I would never point a customer at a 700 line bash script to install a helm chart. 

If we want a CLI to facilitate installing Codesealer (and I really don't think we should, its a huge indicator that this whole thing is way too complex), we should build a proper one.

File: charts/codesealer/templates/core-manager-deployment.yaml [hunk 23]
```diff
@@ -0,0 +1,62 @@
+{{- if .Values.manager.enabled }}
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  labels:
+   {{- include "codesealer.labels" . | nindent 4 }}
+  name: {{ .Values.manager.name }}
+  namespace: {{ .Values.ingress.namespace }}
+spec:
+  replicas: {{ .Values.manager.replicaCount }}
```
Comment (filips) [inline charts/codesealer/templates/core-manager-deployment.yaml]: No reason to have exposed replica config for the manager, as it is inherently non-HA

File: tests/test_install.sh
Comment (tfarinacci) [reply tests/test_install.sh:1]: Agree we don't want a CLI here.  My intent was to provide Adam guidance on what he need to accommodate in the UI.  I use this to test the HELM chart with the rest of the components and was not intended for automated testing.  Until Adam has the UI what should we do with this file?

Comment (psve) [reply tests/test_install.sh:1]: If you want to maintain it for your own purposes, be my guest, but I don't intend on reviewing it.

File: charts/codesealer/templates/core-manager-deployment.yaml (hunk 23 above)
Comment (tfarinacci) [reply charts/codesealer/templates/core-manager-deployment.yaml]: Will hardcode to 1

File: charts/codesealer/values.yaml (hunk 22 above)
Comment (tfarinacci) [reply charts/codesealer/values.yaml]: Removed

File: charts/codesealer/templates/NOTES.txt (hunk 12 above)
Comment (tfarinacci) [reply charts/codesealer/templates/NOTES.txt:9]: I added a line to clarify this.

File: charts/codesealer/README.md (hunk 11 above)
Comment (tfarinacci) [reply charts/codesealer/README.md:68]: Originally we were bypassing the repo install.  This now matches what I have in the test script.

File: img/sidecar-proxy.png
Comment (tfarinacci) [reply img/sidecar-proxy.png:1]: I will create new versions in the next pull request.  Can we proceed for now?

**Review by psve — APPROVED** [review]
Good enough for now
